### PR TITLE
Rationalise discovery of available memory and CPUs

### PR DESCRIPTION
**Pull request by benjaminhwilliams**
At present, there are various different approaches within DIALS for discovering the available memory and CPUs on the system.  This leads to some patchy behaviour in certain environments, most notably when running on a Slurm-managed cluster node, which uses cgroups to establish the memory allocated to the job.  At present, for example, `dials.integrate` will discover a Slurm node's entire memory and try to use it, rather than the memory limit imposed by the scheduler, and so it will frequently be killed by the OOM killer.

Rather than reinvent the wheel and contribute to the proliferation of standards within DIALS for discovery of resources, I've tried here to strip out all the existing memory and CPU discovery methods and replace them with tools cribbed from [Dask](https://www.dask.org/) and [Dask Distributed](https://distributed.dask.org/en/stable/), on the basis that the devs there have thought longer and harder about cross-platform compatibility than we have.  I've also tried to preserve the commit history from those two projects, both to give credit where credit is due and also to provide some history that explains why things are done in this way.

As a side-effect of these changes, I've removed the fall-back option within DIALS integrate to treat swap space as a simple extension of virtual memory and try to allocate to it.  Swapping by design doesn't seem to be a good idea and, in trying to test these changes, I've not actually been able to manufacture a situation in which it actually works — on a RHEL8 workstation with 16GB of RAM and 8GB of swap, the OOM killer still intervenes when `dials.integrate` exceeds the available virtual memory.  And, of course, there's no swap anyway in an HPC context, so it's irrelevant there.

I've tested these changes fairly extensively using a Slurm cluster and a PC with RHEL8 and have no concerns but would welcome other tests on other platforms if people have the time and are willing.

### CODE REVIEW

File: src/dials/algorithms/refinement/refiner.py [hunk 1]
```diff
@@ -483,7 +482,7 @@ def _build_refiner(cls, params, experiments, refman, ref_predictor, do_stills):
             dense_jacobian_gigabytes = (
                 nparam * nref * ndim * flex.double.element_size()
             ) / 1e9
-            avail_memory_gigabytes = psutil.virtual_memory().available / 1e9
+            avail_memory_gigabytes = MEMORY_LIMIT / 1e9
```
Comment (graeme-winter) [inline src/dials/algorithms/refinement/refiner.py:485]: as above, GB or GiB?

File: src/dials/util/system.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations
```
Comment (graeme-winter) [inline src/dials/util/system.py:1]: Not reviewing these changes as assume they were pulled verbatim from the upstream source. 
I trust that the proposal as made is a sensible one.

File: src/dials/algorithms/integration/integrator.py [hunk 3]
```diff
@@ -1283,24 +1280,19 @@ def _run_processor(reflections):
...
                 tables = self.reflections.random_split(
                     self.params.integration.mp.n_subset_split
                 )
             else:
+                # need to do a memory check and decide whether to split table
                 tables = _iterative_table_split(
                     [self.reflections],
                     self.experiments,
-                    available_immediate,
+                    MEMORY_LIMIT * self.params.integration.block.max_memory_usage,
```
[the remark this answers is not in the document]
Comment (graeme-winter) [reply src/dials/algorithms/integration/integrator.py:1295]: It is used as fractional elsewhere - nevertheless the comment would be useful here.

[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply src/dials/algorithms/integration/integrator.py:1295]: Will do.

File: newsfragments/2619.misc [hunk 4]
```diff
@@ -0,0 +1 @@
+Do more accurate discovery of available CPUs and virtual memory across a range of different systems.  This should make various DIALS tools, in particular ``dials.integrate``, function better in situations where system resources are fairly limited or DIALS requires a lot of memory.
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply newsfragments/2619.misc:1]: Sure thing.

File: src/dials/algorithms/refinement/refiner.py (hunk 1 above)
Comment (benjaminhwilliams) [reply src/dials/algorithms/refinement/refiner.py:485]: See above.

File: src/dials/util/system.py (hunk 2 above)
Comment (benjaminhwilliams) [reply src/dials/util/system.py:1]: OK.

**Comment by benjaminhwilliams** [comment]
Hoping that merging main into this branch will have sorted test failures that seem unrelated to changes here.

File: src/dials/algorithms/integration/parallel_integrator.py [hunk 5]
```diff
@@ -300,9 +294,9 @@ def _assert_enough_memory(required_memory, max_memory_usage):
     include increasing the percentage of memory allowed for shoeboxes or
     decreasing the block size. This could also be caused by a highly mosaic
     crystal model - is your crystal really this mosaic?
-      Total system memory: {total_memory / 1000000000.0:.1f} GB
-      Limit image memory: {limit_memory / 1000000000.0:.1f} GB
-      Required image memory: {required_memory / 1000000000.0:.1f} GB
+      Total system memory: {total_memory / 1e9:.1f} GB
```
[the remark this answers is not in the document]
Comment (graeme-winter) [reply src/dials/algorithms/integration/parallel_integrator.py:297]: Agree

**Pull request merged**